### PR TITLE
Fix marker validator with cycler (allow mix of classes) 

### LINKED ISSUE #27596: [Bug]: Markers with numeric name like CARETLEFT cannot be specified using a cycler
### Bug summary

I want to cycle between multiple markerstyles, including those with a numeric value like CARETRIGHT as documented [here](https://matplotlib.org/stable/api/markers_api.html#module-matplotlib.markers), but I get an exception if I add one of those.

### Code for reproduction

```python
#!/usr/bin/env python3

from cycler import cycler
import matplotlib.pyplot as plt
import matplotlib.markers as mk

cycle = cycler(marker=["+",mk.CARETRIGHT])

t=[0,1]
y=[5,6]

plt.rc('axes', prop_cycle=cycle)
plt.plot(t,y)
plt.show()
```

### Actual outcome

```
Traceback (most recent call last):
  File "/home/mcd/bugs/bug-cycler.py", line 11, in <module>
    plt.rc('axes', prop_cycle=cycle)
  File "/usr/lib/python3.11/site-packages/matplotlib/pyplot.py", line 670, in rc
    matplotlib.rc(group, **kwargs)
  File "/usr/lib/python3.11/site-packages/matplotlib/__init__.py", line 1062, in rc
    rcParams[key] = v
    ~~~~~~~~^^^^^
  File "/usr/lib/python3.11/site-packages/matplotlib/__init__.py", line 734, in __setitem__
    raise ValueError(f"Key {key}: {ve}") from None
ValueError: Key axes.prop_cycle: Could not convert 5 to str
```

### Expected outcome

A plot using CARETLEFT as marker.

### Additional information

This seems like it has to do with the fact, that some markers are referred to by number and not by a string like "+".
I don't know whether the problem is in matplotlib, cycler or the combination of both.

### Operating system

Artix Linux

### Matplotlib Version

3.8.1

### Matplotlib Backend

GTK4Agg

### Python version

Python 3.11.6

### Jupyter version

_No response_

### Installation

Linux package manager

### PR DESCRIPTION
**Pull request by faisal-fawad**
## PR summary

Closes #27596. Added basic support for combination of strings and integers to satisfy expected behavior. 

Since this is my first PR, I was unsure on where to include testcases. 

## PR checklist

- [X] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [X] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by ksunden** [comment]
Please add the following to the `rcsetup.pyi` file (Should be in roughly the same order as the `.py` file):

```python
def validate_marker(s: Any) -> int | string: ...
def validate_markerlist(s: Any) -> list[int | string]: ...
```

File: lib/matplotlib/tests/test_cycles.py [hunk 1]
```diff
@@ -27,6 +27,16 @@ def test_marker_cycle():
     assert [l.get_marker() for l in ax.lines] == ['.', '*', 'x', '.']
 
 
+@pytest.mark.parametrize('marker', [
+    ['+', mpl.markers.CARETRIGHT],
+    ['*', '.', 'x'],
+    [1, 2, 3]
```
Comment (ksunden) [inline lib/matplotlib/tests/test_cycles.py:33]:
```suggestion
    [1, 2, 3],
    [1, "+", ".", 4],
```

Just a slightly more explicit test of mixed inputs, as that is the stated goal here.

Technically covered by the first test, but just from looking at the source here was not _sure_ that `CARETRIGHT` is an int and not a string (it _is_ an int, `5`)

**Review by ksunden — APPROVED** [review]
test failures unrelated

File: lib/matplotlib/rcsetup.py [hunk 2]
```diff
@@ -209,6 +209,20 @@ def validator(s):
     validate_float, doc='return a list of floats')
 
 
+def validate_marker(s):
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:212]: This doesn't need to be public.
```suggestion
def _validate_marker(s):
```

File: lib/matplotlib/rcsetup.py [hunk 3]
```diff
@@ -209,6 +209,20 @@ def validator(s):
...
+        return validate_int(s)
+    except ValueError as e:
+        try:
+            return validate_string(s)
+        except ValueError as e:
+            raise ValueError('Supported markers are [string, int]') from e
+
+
+validate_markerlist = _listify_validator(
+    validate_marker, doc='return a list of markers')
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:223]:
```suggestion
_validate_markerlist = _listify_validator(
    _validate_marker, doc='return a list of markers')
```

File: lib/matplotlib/tests/test_cycles.py [hunk 4]
```diff
@@ -27,6 +27,17 @@ def test_marker_cycle():
     assert [l.get_marker() for l in ax.lines] == ['.', '*', 'x', '.']
 
 
+@pytest.mark.parametrize('marker', [
+    ['+', mpl.markers.CARETRIGHT],
+    ['*', '.', 'x'],
+    [1, 2, 3],
+    [1, "+", ".", 4],
+])
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_cycles.py:35]: Is there any particular need to have 4 cycles instead of 1?

**Review by QuLogic — CHANGES_REQUESTED** [review]

File: lib/matplotlib/tests/test_cycles.py (hunk 4 above)
Comment (faisal-fawad) [reply lib/matplotlib/tests/test_cycles.py:35]: Not particularly, the last test case `[1, "+", ".", 4]` should suffice with testing both strings and integers.

**Comment by faisal-fawad** [comment]
@story645 If you could handle it for this PR that would be preferred. 

If you don't mind me asking, how could I squash the commits myself for future PRs?

**Comment by QuLogic** [comment]
There are a few extra changes here; make sure you use [a force push](https://matplotlib.org/devdocs/devel/development_workflow.html#force-push) after rebasing.

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @faisal-fawad! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**